File: R/equiKmeans.R
library(tidyverse)

#' Equal-sized k-means
#' 
#' 
#' @param x numeric matrix of data, or an object that can be coerced to such a
#'   matrix (such as a numeric vector or a data frame with all numeric columns).
#' @param iterations Number of iterations to alternate
#' @param k number of groups to create
#' @param ... further arguments to kmeans()
#' 
#' @details This is an implementation of the equal-sized k-means algorithm published
#'   at https://rviews.rstudio.com/2019/06/13/equal-size-kmeans/
equiKmeans <- function(x, loops, centers) {
  
  NewCenters <- NULL 
  
  for (l in 1:loops) {
    
    # initial kmeans
    if (is.null(NewCenters)) {
      x %>% kmeans(centers = centers, iter.max = 10) -> kclust
      centers = NULL
    } else {
      x %>% kmeans(centers = NewCenters, iter.max = 10)
    }
    
    # compute distance between each center and all points
    kc <- kclust$centers 
    k <- nrow(kc)
    distances <- lapply(1:k, function(i){
      as.matrix(dist(rbind(kc[i, ], x)))[, 1][-1]
    }) %>%
      set_names(str_c("D", 1:k)) %>%
      bind_cols()
    
    # append distances to dataframe
    working <- x %>% bind_cols(distances) %>%
      mutate(assigned = 0, index = 1:n())
    
    kdat <- x %>%
      mutate(assigned = 0, index = 1:n())
    
    FirstRound = nrow(kdat) - (nrow(kdat) %% k) 
    
    # Greedy cluster assignment
    # loop through the points an even number of times
    for(i in 1:FirstRound){ 
      
      #cluster counts can be off by 1 due to uneven multiples of k. 
      j = if(i %% k == 0) k else (i %% k)
      
      # identify the row of the working dataset with the smallest distance
      itemloc = working$index[ 
        which(working[,(paste0("D", j))] == 
                min(working[,(paste0("D",j))]))[1]]
      
      kdat$assigned[kdat$index == itemloc] = j
      
      # remove that row from the working table
      working %<>% filter(!index == itemloc)
      ##The sorting hat says... GRYFFINDOR!!! 
    }
    
    # allocate leftover points to clusters
    if(nrow(working) >= 1){
      for(i in 1:nrow(working)){
        #these leftover points get assigned to whoever's closest, without regard to k
        kdat$assigned[kdat$index ==
                        working$index[i]] = 
          which(working[i,3:5] == min(working[i, 3:5])) 
      }
    }
    
    # Create new centers as means of adjusted cluster memberships
    NewCenters <- kdat %>%
      select(-index)  %>%
      group_by(assigned) %>%
      summarise(across(.cols = everything(), mean)) %>%
      select(-assigned) %>% as.matrix()
  }
  
  list("data" = kdat, "centers" = NewCenters, "kclust" = kclust)
  
}

File: python/equiKmeans.py
import pandas as pd
import numpy as np
from sklearn.cluster import KMeans

def equiKmeans(x, loops, centers, **kwargs):
    """
    This is an implementation of the equal-sized k-means algorithm    
    Parameters:
        x (matrix): such as a numeric vector or a data frame with all numeric columns
        loops (int): Number of iterations to alternate
        center (int): number of groups to create

    Returns:
        : 
    """
    NewCenters = None
    
    for l in range(1, loops + 1):
        # Initial k-means
        if NewCenters is None:
            kclust = KMeans(n_clusters=centers, max_iter=10, **kwargs).fit(x)
            centers = None
        else:
            kclust = KMeans(n_clusters=NewCenters.shape[0], init=NewCenters, max_iter=10, **kwargs).fit(x)
        
        # Compute distance between each center and all points
        kc = kclust.cluster_centers_
        k = kc.shape[0]
        distances = np.zeros((x.shape[0], k))
        
        for i in range(k):
            distances[:, i] = np.linalg.norm(x - kc[i], axis=1)[1:]
        
        distances_df = pd.DataFrame(distances, columns=[f'D{i+1}' for i in range(k)])
        x_with_distances = pd.concat([pd.DataFrame(x), distances_df], axis=1)
        x_with_distances['assigned'] = 0
        x_with_distances['index'] = range(1, x.shape[0] + 1)
        kdat = x_with_distances.copy()
        FirstRound = x.shape[0] - (x.shape[0] % k)
        
        # Greedy cluster assignment
        for i in range(1, FirstRound + 1):
            j = k if i % k == 0 else i % k
            itemloc = x_with_distances['index'][
                x_with_distances[f'D{j}'] == x_with_distances[f'D{j}'].min()].iloc[0]
            kdat.loc[kdat['index'] == itemloc, 'assigned'] = j
            x_with_distances = x_with_distances[x_with_distances['index'] != itemloc]
        
        # Allocate leftover points to clusters
        if len(x_with_distances) >= 1:
            for i in range(len(x_with_distances)):
                k = np.argmin(x_with_distances.iloc[i, 2:2+k])
                kdat.loc[kdat['index'] == x_with_distances.iloc[i, 1], 'assigned'] = k
        
        # Create new centers as means of adjusted cluster memberships
        NewCenters = kdat.drop(columns=['index', 'assigned']).groupby('assigned').mean().values

    return {"data": kdat, "centers": NewCenters, "kclust": kclust}
